File: Dynamic Programming Applications In Machine Learning and Genomics/Probability of a Path in an HMM/HiddenPathProbability.java
import java.util.Scanner;
import java.util.HashMap;

public class CompletedStarter {

    /*
     * Computes the probability of the path pi given an HMM
     *
     * @param pi            a String representing the path of states
     * @param states        an array of chars representing possible states
     * @param transition    a HashMap s.t. transition.get(x).get(y)
     *                      is the probability of transitioning from state
     *                      x to state y. x and y are both chars in states.
     *              
     *
     * @return a double with the correct probability to at least three 
     * significant figures
     */
    public static double pathProbability(String pi, char[] states, 
            HashMap<Character, HashMap<Character, Double>> transition) {
        // TODO: your code here
        return 0.0;
    }

    public static void main(String[] args) {
        Scanner stdin = new Scanner(System.in);
        String pi = stdin.nextLine();

        stdin.nextLine(); // delimiter

        char[] states = String.join("", stdin.nextLine().split(" ")).toCharArray();

        stdin.nextLine(); // delimiter
        stdin.nextLine(); // first line of transition

        HashMap<Character, HashMap<Character, Double>> transition;
        transition = new HashMap<>();

        for (char state : states) {
            HashMap<Character, Double> inner = new HashMap<>();
            String[] currLine = stdin.nextLine().split("\t");

            for (int i = 1; i < currLine.length; i++)
                inner.put(states[i-1], Double.valueOf(currLine[i]));
            
            transition.put(state, inner);
        }

        System.out.println(pathProbability(pi, states, transition));
    }
}

File: Dynamic Programming Applications In Machine Learning and Genomics/Probability of a Path in an HMM/hidden_path_probability.cpp
#include <iostream>
#include <string>
#include <vector>
#include <map>

using std::string;
using std::vector;
using std::map;
using std::cin;
using std::cout;
using std::endl;


/*
 * Computes probability of a path pi through a given HMM.
 *
 * \param pi            a string representing the path
 * \param states        a vector of chars containing all possible states
 * \param transition    a map s.t. transition[x][y] is the probability
 *                      of transitioning from state x to state y. 
 *                      x and y are chars in states.
 *
 * \return a double with the probability to at least three significant figures
 */
double path_probability(const string& pi, const vector<char>& states, 
        const map<char,map<char,double>>& transition) {
    // TODO: your code here
    return 0.0;
}

int main(int argc, char* argv[]) {
    string pi;
    vector<char> states;
    map<char,map<char,double>> transition;

    // read in pi
    cin >> pi;
    cin.ignore(1, '\n'); // last newline
    cin.ignore(100, '\n'); // delimiter

    // read in states
    char state;
    cin >> state;
    while (state != '-') {
        states.push_back(state);
        cin >> state;
    }

    cin.ignore(100, '\n'); // delimiter
    cin.ignore(100, '\n'); // first line of transition matrix

    // read in transition matrix
    char curr_state;
    double curr_prob;
    while (cin >> curr_state) {
        for (char state : states) {
            cin >> curr_prob;
            transition[curr_state][state] = curr_prob;
        }
    }

    cout << path_probability(pi, states, transition) << endl;
    return 0;
}
